File: LongestIncreasingSubsequence.java
/**
 * LeetCode 300 Longest Increasing Subsequence
 */
class Solution {
    public int lengthOfLIS(int[] nums) {
        int maxLength = 0;
        if (nums == null || nums.length == 0) {
            return maxLength;
        }

        int[] dp = new int[nums.length];

        for (int i = 0; i < dp.length; i++) {
            dp[i] = 1;
        }

        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i],dp[j] + 1);
                }
            }
            maxLength = Math.max(maxLength,dp[i]);
        }

        return maxLength;
    }
}

File: LongestIncreasingSubsequence.go
func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}

func lengthOfLIS(nums []int) int {
	n := len(nums)
	dp := make([]int,n)
	maxRes := 0
	for i := 0;i < n;i++ {
		dp[i] = 1
	}
	
	for i := 0;i < n;i++ {
		for j := 0;j < i;j++ {
			if (nums[j] < nums[i]) {
				dp[i] = max(dp[i],dp[j] + 1)
			}
		}
		maxRes = max(maxRes,dp[i])
	}
	return maxRes
}
